runtime/parachains/src/scheduler_common/mod.rs

```diff
@@ -0,0 +1,344 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The scheduler module for parachains and parathreads.
+//!
+//! This module is responsible for two main tasks:
+//!   - Partitioning validators into groups and assigning groups to parachains and parathreads
+//!   - Scheduling parachains and parathreads
+//!
+//! It aims to achieve these tasks with these goals in mind:
+//! - It should be possible to know at least a block ahead-of-time, ideally more,
+//!   which validators are going to be assigned to which parachains.
+//! - Parachains that have a candidate pending availability in this fork of the chain
+//!   should not be assigned.
+//! - Validator assignments should not be gameable. Malicious cartels should not be able to
+//!   manipulate the scheduler to assign themselves as desired.
+//! - High or close to optimal throughput of parachains and parathreads. Work among validator groups should be balanced.
+//!
+//! The Scheduler manages resource allocation using the concept of "Availability Cores".
+//! There will be one availability core for each parachain, and a fixed number of cores
+//! used for multiplexing parathreads. Validators will be partitioned into groups, with the same
+//! number of groups as availability cores. Validator groups will be assigned to different availability cores
+//! over time.
+
+use frame_support::pallet_prelude::*;
+use primitives::{
+	CollatorId, CoreIndex, CoreOccupied, GroupIndex, Id as ParaId, ParathreadClaim,
+	ParathreadEntry, ScheduledCore,
+};
+use scale_info::TypeInfo;
+use sp_std::{collections::btree_map::BTreeMap, prelude::*};
+
+use crate::initializer::SessionChangeNotification;
+
+//#[cfg(test)]
+//mod tests;
+
+/// Reasons a core might be freed
+#[derive(Clone, Copy)]
+pub enum FreedReason {
+	/// The core's work concluded and the parablock assigned to it is considered available.
+	Concluded,
+	/// The core's work timed out.
+	TimedOut,
+}
+
+/// The assignment type.
+#[derive(Clone, Encode, Decode, TypeInfo)]
+#[cfg_attr(feature = "std", derive(PartialEq, Debug))]
+pub enum AssignmentKind {
+	/// A parachain.
+	Parachain,
+	/// A parathread.
+	Parathread(CollatorId, u32),
```

Comment:
The `u32` here is a 'retries' count which is very parathread-specific. Is it possible to encapsulate that in the `CoreAssigner` logic?

Reply:
I would have a preference to track `PeerId` instead of `CollatorId`, because it is available earlier in the connection sequence, thus it is better suited to mitigate sybil attacks: Malicious peers can be disconnected quicker this way. In theory even during the connection setup handshake already. While the `CollatorId` is only found during application level (collator protocol) handshake on top of an already established connection.